File: internal/tools/subscribe.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/Howthemeaning/gnmi-mcp-server/internal/config"
	"github.com/Howthemeaning/gnmi-mcp-server/internal/gnmi"
	"github.com/Howthemeaning/gnmi-mcp-server/internal/session"
	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type SubscribeInput struct {
	ConnParams
	Path              string `json:"path"`
	Mode              string `json:"mode" jsonschema:"ONCE or STREAM"`
	StreamMode        string `json:"stream_mode,omitempty" jsonschema:"SAMPLE/ON_CHANGE/TARGET_DEFINED"`
	SampleInterval    string `json:"sample_interval,omitempty" jsonschema:"e.g. 10s"`
	HeartbeatInterval string `json:"heartbeat_interval,omitempty"`
	SessionName       string `json:"session_name,omitempty"`
	Timeout           int    `json:"timeout,omitempty" jsonschema:"ONCE timeout seconds, default 30"`
}

func parseDur(s string) time.Duration {
	if s == "" {
		return 0
	}
	d, err := time.ParseDuration(s)
	if err != nil {
		return 0
	}
	return d
}

func doSubscribe(ctx context.Context, client gnmi.GnmiClient, cfg *config.AppConfig, mgr *session.Manager, in SubscribeInput) (string, bool) {
	if err := validatePath(in.Path); err != nil {
		return err.Error(), true
	}
	dev, err := in.apply(cfg)
	if err != nil {
		return err.Error(), true
	}
	mode := strings.ToUpper(strings.TrimSpace(in.Mode))
	if mode == "" {
		mode = "ONCE"
	}
	if mode == "POLL" {
		return "POLL mode is not supported; use STREAM for continuous telemetry or ONCE for a one-time snapshot", true
	}
	if mode != "ONCE" && mode != "STREAM" {
		return fmt.Sprintf("unknown mode %q; use ONCE or STREAM", in.Mode), true
	}
	p := gnmi.SubParams{
		Path: in.Path, Mode: mode, StreamMode: in.StreamMode,
		SampleInterval: parseDur(in.SampleInterval), HeartbeatInterval: parseDur(in.HeartbeatInterval),
	}

	if mode == "ONCE" {
		to := in.Timeout
		if to <= 0 {
			to = 30
		}
		octx, cancel := context.WithTimeout(ctx, time.Duration(to)*time.Second)
		defer cancel()
		resp, err := client.SubscribeOnce(octx, dev, p)
		if err != nil {
			return err.Error(), true
		}
		return string(resp), false
	}

	// STREAM: create background session.
	info, err := mgr.Create(ctx, client, dev, p, in.SessionName)
	if err != nil {
		return err.Error(), true
	}
	b, _ := json.Marshal(info)
	return string(b), false
}

func RegisterSubscribe(server *mcp.Server, client gnmi.GnmiClient, cfg *config.AppConfig, mgr *session.Manager) {
	mcp.AddTool(server, &mcp.Tool{
		Name:        "gnmi_subscribe",
		Description: "Subscribe to telemetry. ONCE returns a snapshot; STREAM starts a background session managed via gnmi_session_* (POLL is not supported).",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, in SubscribeInput) (*mcp.CallToolResult, any, error) {
		text, isErr := doSubscribe(ctx, client, cfg, mgr, in)
		return textResult(text, isErr), nil, nil
	})
}

File: src/gnmi_mcp_server/tools/subscribe.py
"""MCP tool: gnmi_subscribe -- subscribe to device telemetry."""

import json
import logging

logger = logging.getLogger(__name__)


def register_subscribe_tool(server, gnmic_client, session_mgr, config):
    """Register the gnmi_subscribe tool."""

    @server.tool()
    async def gnmi_subscribe(
        path: str,
        mode: str,  # ONCE / POLL / STREAM
        target: str = "",
        address: str = "",
        stream_mode: str = "TARGET_DEFINED",
        sample_interval: str = "",
        heartbeat_interval: str = "",
        session_name: str = "",
        timeout: int = 30,
        insecure: bool = False,
        tls_ca: str = "",
        tls_cert: str = "",
        tls_key: str = "",
    ) -> str:
        """Subscribe to gNMI device telemetry data.

        ONCE mode: One-time snapshot, returns immediately.
        STREAM/POLL mode: Starts background collection, use session tools to manage.

        Args:
            path: gNMI subscription path (e.g., /state/port/statistics).
            mode: ONCE (one-shot), POLL, or STREAM (continuous).
            target: Device alias from GNMI_DEVICES config.
            address: Direct host:port (requires GNMI_ALLOW_ARBITRARY=true).
            stream_mode: For STREAM: SAMPLE, ON_CHANGE, or TARGET_DEFINED.
            sample_interval: Sampling interval for SAMPLE mode (e.g., 10s).
            heartbeat_interval: Heartbeat interval (e.g., 30s).
            session_name: Name for STREAM/POLL sessions (auto-generated if empty).
            timeout: Timeout in seconds for ONCE mode (default 30).
            insecure: Skip TLS verification.
            tls_ca: TLS CA certificate path.
            tls_cert: TLS client certificate path.
            tls_key: TLS client key path.
        """
        from gnmi_mcp_server.tools._common import resolve_device, build_extra_args, validate_path

        try:
            validate_path(path)
            device = resolve_device(config, target, address)
        except ValueError as e:
            return str(e)

        mode_upper = mode.upper()
        valid_modes = {"ONCE", "POLL", "STREAM"}
        if mode_upper not in valid_modes:
            return f"Error: mode must be one of: {', '.join(sorted(valid_modes))}"

        if mode_upper == "ONCE":
            extra_args = build_extra_args(device, insecure, tls_ca, tls_cert, tls_key)
            extra_args.extend(["--path", path, "--mode", "once"])
            result = await gnmic_client.run("subscribe", extra_args, device, timeout=timeout)
            if result.is_success:
                return result.stdout
            else:
                return f"Error: {result.error_message}"

        # STREAM / POLL -- start background session
        try:
            session = await session_mgr.create(
                gnmic_client=gnmic_client,
                device=device,
                path=path,
                mode=mode_upper,
                stream_mode=stream_mode.upper() if stream_mode else "TARGET_DEFINED",
                sample_interval=sample_interval,
                heartbeat_interval=heartbeat_interval,
                session_name=session_name,
            )
            return json.dumps({
                "session_name": session.name,
                "pid": session.pid,
                "target": session.target,
                "path": session.path,
                "mode": session.mode,
                "status": session.status,
                "output_file": session.output_file,
                "instruction": (
                    f"Use gnmi_session_tail with session_name='{session.name}' to view data. "
                    f"Use gnmi_session_stop to end the session."
                ),
            }, indent=2)
        except Exception as e:
            return f"Error creating subscribe session: {e}"
